docs/source/blueprints/BP002.rst

```diff
@@ -0,0 +1,220 @@
+            {"type": "package", "package": "lvm"}
+        ]
+
+        print(json.dumps(requirements))
+
+Requirements representation as metadata on test docstring
+---------------------------------------------------------
+
+Test writers may want to add the requirements of a test into de test code. The
+option proposed here allows the use of metadata on test docstrings to represent
+the requirements list.
+
+Below is an example of how to define requirements as metadata on docstrings::
+
+        def test_something(self):
+        '''
+            :avocado requirement: [{"type": "url", "url": "https://cabort.com/cabort.c", "hash": "deadbeefdeadbeef"}]
+            :avocado requirement: [{"type": "vmimage", "distro": "fedora", "version": 31, "arch": "x86_64"}]
+            :avocado requirement: [{"type": "package", "package": "lvm"}]
+        '''
+           <test code>
+
+Requirements files location
+===========================
+
+It may be useful for test writers to define a standard source location for the
+requirements JSON files and the requirements Python executable.
+
+This blueprint defines the default location for those files into a requirements
+directory preceded by the test file name. It is also supported the use of
+sub-directories using the name of a specific test to define requirements for
+that test.
+
+The following file tree is an example of possible use for requirements
+directories for a test::
+
+        cabort.py
+        cabort.py.requirements/
+        ├── CAbort.test_2
+        │   └── requirements.py
+        └── requirements.json
+
+In this case, all the tests on `cabort.py`, except for `CAbort.test_2`, use the
+`requirements.json` file located at `cabort.py.requirements`. The `CAbort.test_2`
+test uses its own `requirements.py` located at `CAbort.test_2` directory inside
+the requirements root directory.
+
+Requirements files command-line parameter
+=========================================
+
+It is also possible to use a command-line parameter to define the location of
+the requirements file. For that, this blueprint defines the parameter
+`--requirements-file` followed by the location of the requirements file. As a
+command-line example, we have::
+
+        avocado --requirements-file requirements.json run passtest.py
+
+.. note:: New Issue: Add the support for `--requirements-file` command-line
+          parameter.
+
+Backward Compatibility
+**********************
+
+The implementation of the requirements resolver, proposed here, affects
+Avocado's behavior related to the tasks executed before a test execution starts.
+
+To make the requirements resolver as flexible as possible, the implementation of
+this blueprint may change the utility APIs related to a requirement type.
+
+Security Implications
+*********************
+
+Avocado users should have the warranty that their jobs are running in an isolated
+environment.
+
+We should consider this and keep in mind that any moves here should continue
```

Comment:
It is important to discuss/define what will be the behavior regarding the execution of the `requirements.py` like files, since that has security implications.

Reply:
As far as I could think, it has the same implication of running a malicious test. I don't believe this is under the Avocado's control, or I'm missing something here. Do you have any other ideas of how we could control, or even if we should control? What else are you considering we should discuss here? I'm out of imagination :smiley:

Reply:
The point I was thinking about is that by default, Avocado will not execute test code *unless* told to do so, and shockingly this is different from various test tools.  So, right now it's possible to list test tags with:

```
avocado list -V
```

And only valid Avocado tests will be shown, and their tags should be properly captured and shown.  But, if we do something similar to requirements, say:

```
avocado list -V --also-list-requirements
```

It poses the question whether Avocado should run the `requirements.py`-like files and show those results.  And keep in mind that the test resolution may happen on one environment (say the host machine) and the execution happen elsewhere (say on a container or vm).

Reply:
The ability to list the requirements in different environments when using the `requirements.py` is an interesting problem. I'll be honest here saying I don't see a simple solution right now besides deciding whether we will list by default the requirements from a `requirements.py` file, executing it, or not.

To implement that, we may have a security flag disabling the execution of the `requirements.py` file by default, letting the user decide to enable it or not.

And when the flag is enabled, we may have a **big** warning saying the requirements listed on that output refer to the `list` command executed on that environment and may differ if the test runs on a different environment.

Reply:
That's exactly what I was expecting this would lead.  We should consider those implications and state from the start where we position Avocado.

FIY, this is/should already be the norm in Avocado, with `etc/avocado/conf.d/glib.conf` setting an example:

```
[plugins.glib]
# Wether to run executables in an attempt to discover tests
unsafe = False
```

Reply:
Perfect, thanks for the example!